src/core/experimental/array.d

```diff
@@ -0,0 +1,2559 @@
+        }
+
+        {
+            auto a = immutable Array!(int)(1, 2, 3);
+            auto a2 = a.idup();
+            static assert (is(typeof(a2) == immutable));
+        }
+    }
+
+    /**
+     * Perform a copy of the array. This will create a new array that will copy
+     * the elements of the current array. This will `NOT` call `dup` on the
+     * elements of the array, regardless if `T` defines it or not.
+     *
+     * Returns:
+     *      a new mutable array.
+     *
+     * Complexity: $(BIGOH n).
+     */
+    Array!T dup(this Q)()
+    {
+        Array!T result;
+
+        static if (is(Q == immutable) || is(Q == const))
+        {
+            result.reserve(length);
+            foreach(i; 0 .. length)
+            {
+                result ~= this[i];
+            }
+        }
+        else
+        {
+            result.insert(0, this);
+        }
+        return result;
+    }
+
+    ///
+    static if (is(T == int))
+    @safe unittest
+    {
+        auto stuff = [1, 2, 3];
+        auto a = immutable Array!int(stuff);
+        auto aDup = a.dup;
+        assert(equal(aDup, stuff));
+        aDup.front = 0;
+        assert(aDup.front == 0);
+        assert(a.front == 1);
+    }
+
+    /**
+     * Return a slice to the current array. This is equivalent to calling
+     * `save`.
+     *
+     * Returns:
+     *      an array that references the current array.
+     *
+     * Complexity: $(BIGOH 1)
+     */
+    Qualified opSlice(this Qualified)()
+    {
+        return this.save;
+    }
+
+    /**
+     * Return a slice to the current array that is bounded by `start` and `end`.
+     * `start` must be less than or equal to `end` and `end` must be less than
+     * or equal to `length`.
+     *
+     * Returns:
+     *      an array that references the current array.
+     *
+     * Params:
+     *      start = a positive integer
+     *      end = a positive integer
+     *
+     * Complexity: $(BIGOH 1)
+     */
+    Qualified opSlice(this Qualified)(size_t start, size_t end)
+    in
+    {
+        assert(start <= end && end <= length,
+               "Array.opSlice(s, e): Invalid bounds: Ensure start <= end <= length");
+    }
+    body
+    {
+        return typeof(this)(_support, _payload[start .. end], _isShared);
+    }
+
+    ///
+    static if (is(T == int))
+    @safe unittest
+    {
+        auto stuff = [1, 2, 3];
+        auto a = Array!int(stuff);
+        assert(equal(a[], stuff));
+        assert(equal(a[1 .. $], stuff[1 .. $]));
+    }
+
+    /**
+     * Provide access to the element at `idx` in the array.
+     * `idx` must be less than `length`.
+     *
+     * Returns:
+     *      a reference to the element found at `idx`.
+     *
+     * Params:
+     *      idx = a positive integer
+     *
+     * Complexity: $(BIGOH 1).
+     */
+    ref auto opIndex(this _)(size_t idx)
+    in
+    {
+        assert(idx < length, "Array.opIndex: Index out of bounds");
```

Comment:
This does not behave the same as a built in array.

Reply:
This should also respect `version(D_NoBoundsCheck)`.

Reply:
> Will change. According to the docs, this should throw a RangeError. Am I right?

Yes, but it depends on which compiler setting is used. Have a look at the `boundscheck` flag https://dlang.org/dmd-osx.html#switch-boundscheck

Reply:
> > Will change. According to the docs, this should throw a RangeError. Am I right?
> 
> Yes, but it depends on which compiler setting is used and if it’s a `@safe` function or not. Have a look at the `boundscheck` flag https://dlang.org/dmd-osx.html#switch-boundscheck

@jacob-carlborg do you happen to know how can I check if `boundscheck` is set to `safeonly`? The only conditional compilation I could find was `version (D_NoBoundsCheck)`, so I can distinguish the `on/off` behaviour, but not the `safeonly` one.

Also, RangeError is not `@nogc`, so I've kept the `assert(...)` instead of throwing a RangeError. Any suggestions here?

Reply:
This is solved now, as it's delegating to the underlying built-in array type.